[libc++] Fix assignment in insertion into `vector`

frederick-vs-ja opened the pull request:
Changes:
- Avoid direct assignment in iterator-pair `insert` overload and `insert_range`, except when the assignment is move assignment.

Fixes #47963. Fixes #115801.

philnik777 commented:
Please split adding `std::move` from any algorithm changes.

frederick-vs-ja commented:
> Please split adding `std::move` from any algorithm changes.

I think I should split out the handling for input-only iterators, which is more than adding `std::move`.

philnik777 commented:
> > Please split adding `std::move` from any algorithm changes.
> 
> I think I should split out the handling for input-only iterators, which is more than adding `std::move`.

Sure. Which way you split is up to you. I only care that it is split.

frederick-vs-ja commented:
> > > Please split adding `std::move` from any algorithm changes.
> > 
> > 
> > I think I should split out the handling for input-only iterators, which is more than adding `std::move`.
> 
> Sure. Which way you split is up to you. I only care that it is split.

Split to #116157. I sadly realized why I placed the changes in one PR - `__insert_with_size` needs to be fixed in both directions, and the changes are far from orthogonal.

frederick-vs-ja committed:
[libc++] Fix assignment in insertion into `vector`

Changes:
- Avoid direct assignment in iterator-pair `insert` overload and
`insert_range`, except when the assignment is move assignment.

winner245 commented:
Could you please rebase this PR onto main? Since the split part (#116157) has already been merged, combining them now would ease the review process for this one. Thank you!

winner245 commented on libcxx/include/__vector/vector.h:
```diff
@@ -594,6 +598,30 @@ class _LIBCPP_TEMPLATE_VIS vector {
+            __enable_if_t<!is_same<decltype(*std::declval<_Iterator&>())&&, value_type&&>::value, int> = 0>
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void
+  __insert_assign_n_unchecked(_Iterator __first, difference_type __n, pointer __position) {
+    for (pointer __end_position = __position + __n; __position != __end_position; (void)++__position, ++__first) {
```
```suggestion
    for (pointer __end_position = __position + __n; __position != __end_position; ++__position, (void)++__first) {
```

I think it is more appropriate to guard against `operator,` here.

frederick-vs-ja replied:
Thanks! Although I believe the effects are exactly same.

It's conventional to `void`-cast the first operand in the standard wording. But in libc++ `void`-casting the second one seems conventional. What an unfortunate style mismatch.

frederick-vs-ja committed:
Clang-format `vector.modifiers/insert_iter_iter_iter.pass.cpp`

frederick-vs-ja committed:
Adopt @winner245's review comments

frederick-vs-ja committed:
Fix `this->__alloc()`

frederick-vs-ja committed:
XFAIL: FROZEN-CXX03-HEADERS-FIXME

frederick-vs-ja committed:
Merge branch 'main' into fix-vector-insertion

winner245 commented on libcxx/include/__vector/vector.h:
```diff
@@ -602,6 +603,30 @@ class _LIBCPP_TEMPLATE_VIS vector {
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void
+  __insert_assign_n_unchecked(_Iterator __first, difference_type __n, pointer __position) {
+#if _LIBCPP_STD_VER >= 23
+    if constexpr (!forward_iterator<_Iterator>) {
```
```suggestion
    if constexpr (!forward_iterator<_Iterator>) { // Handles input-only sized ranges for insert_range
```

frederick-vs-ja replied:
Added.

winner245 commented on libcxx/test/std/containers/sequences/vector/vector.modifiers/insert_iter_iter_iter.pass.cpp:
```diff
@@ -161,6 +163,24 @@ TEST_CONSTEXPR_CXX20 bool tests() {
     for (; j < 105; ++j)
       assert(v[j] == 0);
   }
+  {
```
```suggestion
  { // Ensure that iterator-pair insert() doesn't use unexpected assignment.
```

winner245 reviewed (approved):
LGTM. Thanks for the patch. It ensures that our `insert_range` function, as well as the iterator-pair `insert` function, will not accidentally use copy-assignment. I only have a few minor comments to improve readability.

frederick-vs-ja committed:
Merge branch 'main' into fix-vector-insertion

frederick-vs-ja committed:
Adopt @winner245's review comments on comments

frederick-vs-ja committed:
Merge branch 'main' into fix-vector-insertion
